fdbserver/IPager.h

```diff
@@ -97,67 +151,341 @@ class ArenaPage : public ReferenceCounted<ArenaPage>, public FastAllocated<Arena
+			}
+		}
+		void decode(uint8_t secret, uint8_t* payload, int len, PhysicalPageID seed) {
+			for (int i = 0; i < len; ++i) {
+				payload[i] ^= secret;
+			}
+			if (checksum != XXH3_64bits_withSeed(payload, len, seed)) {
+				throw page_decoding_failed();
+			}
+		}
+	};
+#pragma pack(pop)
+
+	// Get the size of the encoding header based on type
+	// Note that this is only to be used in operations involving new pages to calculate the payload offset.  For
+	// existing pages, the payload offset is stored in the page.
+	static int encodingHeaderSize(EncodingType t) {
+		if (t == EncodingType::XXHash64) {
+			return sizeof(XXHashEncodingHeader);
+		} else if (t == EncodingType::XOREncryption) {
+			return sizeof(XOREncryptionEncodingHeader);
+		} else {
+			throw page_encoding_not_supported();
+		}
+	}
 
-	typedef XXH64_hash_t Checksum;
+	// Get the usable size for a new page of pageSize using HEADER_WRITE_VERSION with encoding type t
+	static int getUsableSize(int pageSize, EncodingType t) {
+		return pageSize - sizeof(Page) - sizeof(HeaderV1) - encodingHeaderSize(t);
+	}
 
-	// Usable size, without checksum
-	int size() const { return logicalSize - sizeof(Checksum); }
+	// Initialize the header for a new page so that the payload can be written to
+	// Pre:  Buffer is allocated and logical size is set
+	// Post: Page header is initialized and space is reserved for subheaders for
+	//       HEADER_WRITE_VERSION main header and the given encoding type.
+	//       Payload can be written to with mutateData() and dataSize()
+	void init(EncodingType t, PageType pageType, uint8_t pageSubType) {
+		page->headerVersion = HEADER_WRITE_VERSION;
+		page->encodingHeaderOffset = sizeof(Page) + sizeof(HeaderV1);
+		page->encodingType = t;
+		page->payloadOffset = page->encodingHeaderOffset + encodingHeaderSize(t);
+
+		pPayload = page->getPayload();
+		payloadSize = logicalSize - (pPayload - buffer);
+
+		HeaderV1* h = page->getMainHeader<HeaderV1>();
+		h->pageType = pageType;
+		h->pageSubType = pageSubType;
+
+		// Write dummy values for these in new pages. They should be updated when possible before calling preWrite()
+		// when modifying existing pages
+		h->lastKnownLogicalPageID = invalidLogicalPageID;
+		h->lastKnownParentLogicalPageID = invalidLogicalPageID;
+		h->writeVersion = invalidVersion;
+	}
 
-	Standalone<StringRef> asStringRef() const { return Standalone<StringRef>(StringRef(begin(), size()), arena); }
+	// Get the logical page buffer as a StringRef
+	Standalone<StringRef> asStringRef() const { return Standalone<StringRef>(StringRef(buffer, logicalSize)); }
 
 	// Get an ArenaPage which is a copy of this page, in its own Arena
-	Reference<ArenaPage> cloneContents() const {
+	Reference<ArenaPage> clone() const {
 		ArenaPage* p = new ArenaPage(logicalSize, bufferSize);
 		memcpy(p->buffer, buffer, logicalSize);
+
+		// Non-verifying header parse just to initialize component pointers
+		p->postRead1(invalidPhysicalPageID, false);
```

Comment:
Why "verify" is set to false? Also, do you think making it a KNOB would be more flexible?

Reply:
Non-verifying refers to the `false` parameter, the comment block for `postRead1()` explains what it does.  I'm not sure what else to put here, the comment is just to explain which mode of the function is being used.

Reply:
What I meant to ask was: "why to explicitly NOT verify the content"? But, reading your comment makes sense to me. Thanks.